File: CS302 Modulation and Simulation/Lab8/q1.m
% Given parameters
a = 0.0106;
j = 0.0544;
J0 = 18274;
A0 = 66454;
delta_t = 1; % Time step in days

% Initial conditions
J = J0;
A = A0;
t = 0;

% Lists to store values for plotting
time_list = [];
J_list = [];
A_list = [];

% Euler's method integration
while (J >= 1 && A >= 1)
    % Store current values
    time_list(end+1) = t;
    J_list(end+1) = J;
    A_list(end+1) = A;
    
    % Update equations
    J_dot = -a * A;
    A_dot = -j * J;
    
    J = J + J_dot * delta_t;
    A = A + A_dot * delta_t;
    
    % Increment time
    t = t + delta_t;
end

% Outcome of the battle
if J < 1
    disp('Japanese army defeated.');
    victorious_army = 'American';
    casualties = J0 - J;
    active_troops = A;
elseif A < 1
    disp('American army defeated.');
    victorious_army = 'Japanese';
    casualties = A0 - A;
    active_troops = J;
end

disp(['Victorious army: ' victorious_army]);
disp(['Number of active troops in victorious army: ' num2str(active_troops)]);
disp(['Number of casualties in defeated army: ' num2str(casualties)]);
disp(['Number of days for outcome: ' num2str(t)]);

% Plotting
figure;

% Plot of J and A along with time
%subplot(2,1,1);
plot(time_list, J_list, 'r', 'LineWidth', 2);
hold on;
plot(time_list, A_list, 'b', 'LineWidth', 2);
xlabel('Time (days)');
ylabel('Number of troops');
legend('Japanese', 'American');
title('Number of troops vs Time');
figure;
% Plot with logarithmic vertical axis
%subplot(2,1,2);
semilogy(time_list, J_list, 'r', 'LineWidth', 2);
hold on;
semilogy(time_list, A_list, 'b', 'LineWidth', 2);
xlabel('Time (days)');
ylabel('Number of troops (log scale)');
legend('Japanese', 'American');
xlim([0 30.36]);
title('Number of troops vs Time (log scale)');

% Plot A vs J
figure;
plot(J_list, A_list, 'k', 'LineWidth', 2);
xlabel('Number of Japanese troops');
ylabel('Number of American troops');
title('Number of American troops vs Number of Japanese troops');

File: CS302 Modulation and Simulation/Lab8/Q1.py
import matplotlib.pyplot as plt
import numpy as np

# Define parameters
a = 0.0106
j = 0.0544
J0 = 18274
A0 = 66454
dt = 1  # Time step (days)

# Initialize lists for storing troop numbers
J = [J0]
A = [A0]
t = [0]

# Stopping condition (one army falls below 1 troop)
while min(J[-1], A[-1]) > 1:
    # Update troop numbers using Euler's method
    J_new = J[-1] - a * A[-1] * dt
    A_new = A[-1] - j * J[-1] * dt

    # Append to lists
    J.append(J_new)
    A.append(A_new)
    t.append(t[-1] + dt)

# Calculate number of days and victorious army
days = len(t) - 1
victorious_army = "American" if A[-1] > J[-1] else "Japanese"  # Assuming victory for remaining army

# Calculate casualties (difference between initial and final troop numbers)
american_casualties = A0 - A[-1]
japanese_casualties = J0 - J[-1]

# Significance of casualty numbers
#The high casualty numbers on both sides highlight the ferocity of the Battle of Iwo Jima. Despite the American victory, the losses were significant, underscoring the human cost of war.

# Plot J and A with linear and logarithmic y-axes
plt.figure(figsize=(10, 6))

plt.subplot(121)
plt.plot(t, J, label='Japanese Troops')
plt.plot(t, A, label='American Troops')
plt.xlabel('Time (days)')
plt.ylabel('Troop Count')
plt.title('Linear Scale')
plt.legend()

plt.subplot(122)
plt.semilogy(t, J, label='Japanese Troops')
plt.semilogy(t, A, label='American Troops')
plt.xlabel('Time (days)')
plt.ylabel('Troop Count (Logarithmic)')
plt.title('Logarithmic Scale')
plt.legend()

plt.tight_layout()
plt.show()
